File: C/Sorting/selection_Sort.c
// O(n^2) Time Complexity | O(1) Space Complexity

#include<stdio.h>

#define N 8

//Function Signature
int * selectionSort(int*); //Function to implement selection sort algorithm
void swap(int*, int*); //Function to swap two elements

int main(){
    int myArray[N] = {1, -12, 9, 54, -11, 75, 8, 14};
    int i = 0;
    int *ptr;

    ptr = selectionSort(myArray);
    for(i = 0; i < N; i++){
        printf(" %d ", *(ptr + i));
    }
    return 0;
}

int * selectionSort(int *myArray){
    int currentIdx = 0;
    int lenArray = N;
    int smallestIdx = 0;
    int i = 0;

    while(currentIdx < lenArray - 1){
        smallestIdx = currentIdx;
        for(i  = currentIdx + 1; i < lenArray; i++){
            if(*(myArray + smallestIdx) > *(myArray + i)){
                smallestIdx = i; //Update smaller index
            }
        }
        swap(&*(myArray + currentIdx), &*(myArray + smallestIdx)); //Swap elements to their right position
        currentIdx++; 
    }
    return myArray;
}

void swap(int *i, int *j){
    int temp =0; 

    temp = *i;
    *i = *j;
    *j = temp;
}

File: Python/Sorting/selection_sort.py
# O(n^2) Time Complexity | O(1) Space Complexity

def selection_sort(array):
    """Function to implement selection sort algorithm"""
    current_idx = 0 # Starting index
    while current_idx < len(array) - 1:
        smallest_idx = current_idx
        for i in range(current_idx + 1, len(array)):
            # Serch for smallest number
            if array[smallest_idx] > array[i]:
                smallest_idx = i # Updating smallest number
        swap(current_idx, smallest_idx, array)
        current_idx += 1 # New current index
    return array

def swap(i, j, array):
    """Function to swap two elements"""
    array[i], array[j] = array[j], array[i]

# Main Program
print(selection_sort([1, -12, 9, 54, -11, 75, 8, 14]))